### ISSUE TITLE
cx_Oracle.DatabaseError: DPI-1002: invalid OCI handle with threaded connections

### ISSUE THREAD

**Issue by aguinane** [oracle/python-cx_Oracle#117]
1. What is your version of Python? Is it 32-bit or 64-bit?
Python 3.6.3 (v3.6.3:2c5fed8, Oct  3 2017, 17:26:49) [MSC v.1900 32 bit (Intel)] on win32

2. What is your version of cx_Oracle?
cx-Oracle==6.0.3
Also had the issue on 6.0.2 Have no issues on 5.3 and earlier.

3. What is your version of the Oracle client (e.g. Instant Client)?  How was it installed? Where is it installed?
From path:
C:\Oracle\instantclient_12_1;C:\Oracle\instantclient_11_2;C:\Oracle\instantclient_11_2_64;C:\Oracle\Ora11G\bin;C:\Oracle\112\bin;C:\oracle\ora102\bin;C:\Oracle\Ora92\bin;

It's a work machine with a mess of different Oracle database versions so has lots of client versions.
Have also tried on other work machines with different install combinations with same error.

4. What is your version of the Oracle Database?
Tested on multiple Oracle DBs: version 10.2.0.3.0 and 11.2.0.4.0

5. What is your OS and version?
Windows 7 64 bit

6. What compiler version did you use?  For example, with GCC, run
Installed from python wheel. (via pip)

7. What environment variables did you set?  How *exactly* did you set them?
TNS_ADMIN: C:\Oracle\Ora92\network\ADMIN\

8. What *exact* command caused the problem (e.g. what command did you try to
   install with)?  Who were you logged in as?
```python
import cx_Oracle
constring = 'user/pass@db' 
for i in range(7):
    print('Connection', i)
    con = cx_Oracle.connect(constring, threaded=True)
    cur = con.cursor()
    res = cur.execute('SELECT USER FROM DUAL')
    print('Connected')
print('DONE!')
```

9. What error(s) you are seeing?
Causes an error when connecting to 3rd database in thread mode.
```
Connection 0
Connected
Connection 1
Connected
Connection 2
Traceback (most recent call last):
  File "test_oracle.py", line 8, in <module>
    res = cur.execute('SELECT USER FROM DUAL')
cx_Oracle.DatabaseError: DPI-1002: invalid OCI handle
```
I am using sqlalchemy which calls with thread mode on by default, and originally [raised an issue there](https://bitbucket.org/zzzeek/sqlalchemy/issues/4105/cx_oracle-602-crashes-python-on-6th-db). It appears to fail on the 6th connection when using alchemy settings.

No issues running the code with cx_Oracle 5.3 or with the threaded setting set to false when connecting.

When running via sqlalchemy it actually completely crashes python.

**Comment by anthony-tuininga** [comment]
I just tried your code and do not get the errors you are getting. I suspect configuration is to blame. Can you try on a "clean" machine with only one instant client installed? Can you also set the environment variable DPI_DEBUG_LEVEL=8 before running the code and report the information that is displayed? That might help track things down further.

**Comment by aguinane** [comment]
Unfortunately we have fairly locked down work machines, so can't easily test a clean machine. I have got the same error on some VMs, but they are generally installed the same way. Here is the results with the DPI_DEBUG_LEVEL setting.
```
ODPI: internal error DPI-1002: invalid OCI handle (dpiConn_prepareStmt / acquire mutex)
Traceback (most recent call last):
  File "test_oracle.py", line 8, in <module>
    res = cur.execute('SELECT USER FROM DUAL')
cx_Oracle.DatabaseError: DPI-1002: invalid OCI handle
ODPI: internal error DPI-1002: invalid OCI handle (dpiConn_release / acquire mutex)
```

**Comment by anthony-tuininga** [comment]
Hmm, that code has been removed in the master branch (what will become 6.1). Can you try with that version and see if you still get the same results?

**Comment by aguinane** [comment]
I'm unable to install a compiler to build on windows unfortunately. 
Happy for you to close and I will re-open if it's still an issue when 6.1 comes out - unless someone else can replicate and test in the meantime.

**Comment by anthony-tuininga** [comment]
Ok. I'll close it for now and you can re-open if 6.1 does not address the issue.

**Comment by aguinane** [comment]
This is resolved in 6.1. Thanks :)
